### DISCUSSION TITLE
Using SU2's preaccumulation to obtain a mean variable

### DISCUSSION THREAD

**Discussion by JuanSembla** [su2code/SU2#1760] [Q&A]
Hello all! I hope your're doing well.

I'm trying to add a turbulence model correction to Menter's SST that works in a very similar manner to Spalar-Allmaras' QCR but uses the specific dissipation rate as a parameter (as it is based on Pope's 1975 work). Thus, I'm coding the model on flow_diffusion.cpp but I can't seem to find a way to bring omega into it. I see that SU2 uses pre-accumulation to store variables and bring them when in need, for example, in that same script we have:

AD::StartPreacc();
AD::SetPreaccIn(V_i, nDim+9); AD::SetPreaccIn(V_j, nDim+9);
AD::SetPreaccIn(Coord_i, nDim); AD::SetPreaccIn(Coord_j, nDim);
AD::SetPreaccIn(PrimVar_Grad_i, nDim+1, nDim);
AD::SetPreaccIn(PrimVar_Grad_j, nDim+1, nDim);
AD::SetPreaccIn(turb_ke_i); AD::SetPreaccIn(turb_ke_j);
AD::SetPreaccIn(TauWall_i); AD::SetPreaccIn(TauWall_j);
AD::SetPreaccIn(Normal, nDim);

As far as I understand, this feature is used for algorithmic differentiation. However, in flow_diffusion.cpp it is also used to obtain mean variables as in:

Mean_turb_ke = 0.5*(turb_ke_i + turb_ke_j);

Therefore, think I should achieve something like:

AD::SetPreaccIn(omega_i); AD::SetPreaccIn(omega_j);

In order to do Mean_omega = 0.5*(omega_i + omega_j), which would be what I need to make use of my model. However, I do not seem to be able to store the value (or the address on which it is stored) of omega_i and omega_j so as to bring them in the flow_diffusion.cpp script. Any ideas?

Thank you in advance, I hope you have a nice day.

**Comment by pcarruscag** [comment]
Hi, preaccumulation is for something else.
Have a look at CEulerSolver::Upwind_Residual, the solver sets the values of the numerics class (numerics->Set??(value))

**Reply by JuanSembla** [reply]
Thank you very much for your answer. In fact I was working on CEulerSolver::Upwind_Residual but when I set the value of, for example, Dissipation_i and Dissipation_j using numerics->SetDissipation(diss_i, diss_j), I cannot seem to retrieve the value in flow_diffusion.cpp (in particular I'm working with CAvgGrad_Flow::ComputeResidual). Indeed, if I do cout to see the value that is stored in the variable Dissipation_i it is not the one I set in CEulerSolver::Upwind_Residual, which makes sense because when I look at the memory addresses, they are not the same.
In summary, when setting the value of Dissipation_i and j in CEulerSolver::Upwind_Residual using the Set function it seems as if I were storing the value somewhere that is not the same as the one CAvgGrad_Flow::ComputeResidual searches for it.
Forgive my ignorance in c++, I hope my problem is clear and I thank you in advance for your help, I really appreciate it.

**Reply by pcarruscag** [reply]
Did you add those variables to CNumerics or did they exist already? If they existed it is possible that their values are being modified by the existing code, i.e. they are intermediate instead of input.
Another option is that you are setting the values in the wrong numerics instance. For viscous fluxes you need to set the VISC_TERM instead of CONV_TERM numerics, look for uses of VISC_TERM or visc_numerics in CEulerSolver CNSSolver or CFlowSolverBase

**Reply by JuanSembla** [reply]
I've actually tried both, working with existing variables and adding some too. As to the former, I believe to be using the correct numerics since I'm mimicking the code but it does not seem to be working since when I output the memory address of the variable, it is different from the one my function in  CAvgGrad_Flow::ComputeResidual looks for it. Regarding my attempt to create another set of variables (and a Set..() function too) it did not work either, maybe because I'm not setting the values in the numerics instance correctly. Is there a way I can make sure my function in CAvgGrad_Flow::ComputeResidual looks for the variable in the correct address?
Thank you in advance.
